Add Pastebin support

mcornick opened the pull request:
The Pastebin suite of API calls is very similar to the PURL suite, so the Paste code is based heavily on the PURL code.

mcornick committed:
Add Pastebin support

Signed-off-by: Mark Cornick <mcornick@mcornick.com>

ejstreet commented on omglol/models.go:
```diff
@@ -145,6 +145,12 @@ type dnsChangeResponse struct {
+type Paste struct {
+	Title      string `json:"title"`
+	Content    string `json:"content"`
+	ModifiedOn int64  `json:"modified_on"`
```
I noticed in the API (not yet documented), that there is also now a `listed` attribute, much like with PURLs.
```suggestion
	ModifiedOn int64  `json:"modified_on"`
	Listed     bool   `json:"listed"`
```

ejstreet commented on omglol/paste.go:
```diff
@@ -0,0 +1,166 @@
+			Paste   struct {
+				Title      string `json:"title"`
+				Content    string `json:"content"`
+				ModifiedOn int64  `json:"modified_on"`
```
```suggestion
				ModifiedOn int64  `json:"modified_on"`
				Listed     *int64 `json:"listed"`
```

ejstreet commented on omglol/paste.go:
```diff
@@ -0,0 +1,166 @@
+	return &Paste{
+		Title:      Title,
+		Content:    Content,
+		ModifiedOn: ModifiedOn,
```
`ModifiedOn` will be added by the API, so make it optional when used for `CreatePaste`. Also add the `listed` attribute.
```suggestion
	func NewPaste(Title, Content string, Listed bool, ModifiedOn ...int64) *Paste {
	var modified *int64
	if len(ModifiedOn) > 0 {
		m := ModifiedOn[0]
		modified = &m
	}
	
	return &Paste{
		Title:      Title,
		Content:    Content,
		Listed:     Listed,
		ModifiedOn: modified,
```

ejstreet commented on omglol/paste.go:
```diff
@@ -0,0 +1,166 @@
+
+// Create a new Paste. See https://api.omg.lol/#token-post-pastes-create-a-new-paste
+func (c *Client) CreatePaste(domain string, paste Paste) error {
+	type pasteRequest struct {
+		Title   string `json:"title"`
+		Content string `json:"content"`
+	}
+
+	p := pasteRequest{
+		Title:   paste.Title,
+		Content: paste.Content,
+	}
+
```
Add in `listed` just like in PURLs.
```suggestion
	type pasteRequest struct {
		Title   string `json:"title"`
		Content string `json:"content"`
		Listed  *bool  `json:"listed"`
	}

	p := pasteRequest{
		Title:   paste.Title,
		Content: paste.Content,
	}
	
	if !paste.Listed {
		p.Listed = nil
	} else {
		t := true
		p.Listed = &t
	}

```

ejstreet commented on omglol/paste.go:
```diff
@@ -0,0 +1,166 @@
+		return nil, err
+	}
+
+	return NewPaste(g.Response.Paste.Title, g.Response.Paste.Content, g.Response.Paste.ModifiedOn), nil
```
Add in `listed` while handling API quirks.
```suggestion
	var listed bool
	if g.Response.Paste.Listed != nil {
		listed = true
	} else {
		listed = false
	}
	
	return NewPaste(g.Response.Paste.Title, g.Response.Paste.Content, listed, g.Response.Paste.ModifiedOn), nil
```

ejstreet commented on omglol/paste.go:
```diff
@@ -0,0 +1,166 @@
+		} `json:"request"`
+		Response struct {
+			Message  string `json:"message"`
+			Pastebin []struct {
+				Title      string `json:"title"`
+				Content    string `json:"content"`
+				ModifiedOn int64  `json:"modified_on"`
+			} `json:"pastebin"`
```
Better to use the existing `Paste` type I think - I realise I should update this in PURLs also.
```suggestion
			Pastebin []Paste `json:"pastebin"`
```

ejstreet commented on omglol/paste.go:
```diff
@@ -0,0 +1,166 @@
+
+	body, err := c.doRequest(req)
+	if err != nil {
+		if strings.Contains(err.Error(), "status: 404") {
```
```suggestion
		// Return an empty list instead of erroring if no pastes exist
		if strings.Contains(err.Error(), "status: 404") {
```

ejstreet commented on omglol/paste.go:
```diff
@@ -0,0 +1,166 @@
+		x.Title = paste.Title
+		x.Content = paste.Content
+		x.ModifiedOn = paste.ModifiedOn
+		p = append(p, x)
```
This loop should probably be a shared function with GetPaste, but I'll tidy that up some other time.

Add `listed`.
```suggestion
		if paste.Listed == nil {
			x.Listed = false
		} else {
			x.Listed = true
		}
		
		p = append(p, x)
```

ejstreet commented on omglol/paste_test.go:
```diff
@@ -0,0 +1,90 @@
+	if len(p.Title) <= 0 {
+		t.Error("Paste Title is empty.")
+	}
+	if len(p.Content) <= 0 {
+		t.Error("Paste Content is empty.")
+	}
```
An empty paste is actually valid. We can check to make sure that the timestamp is being set though. 
```suggestion
	if p.ModifiedOn <= 0 {
		t.Error("ModifiedOn time is unset.")
	}
```

ejstreet commented on omglol/paste.go:
```diff
@@ -0,0 +1,166 @@
+
+// Returns a string representaion of a Paste
+func (p *Paste) String() string {
+	return fmt.Sprintf("Title: %s, Content: %s, ModifiedOn: %d", p.Title, p.Content, p.ModifiedOn)
```
```suggestion
	return fmt.Sprintf("Title: %s, Content: %s, Listed: %t, ModifiedOn: %d", p.Title, p.Content, p.Listed, p.ModifiedOn)
```

ejstreet commented on omglol/paste_test.go:
```diff
@@ -0,0 +1,90 @@
+
+	title := "test" + RunUID
+
+	paste := NewPaste(title, "example paste content", 0)
```
```suggestion
	unlistedPaste := NewPaste(unlistedTitle, "example paste content", true)
```

ejstreet commented on omglol/paste_test.go:
```diff
@@ -0,0 +1,90 @@
+
+	if u != nil {
+		t.Log(u.String())
+		validatePaste(t, *u)
```
Make sure that our bool/int/null wrangling has correctly set Listed. This half of the test isn't that great because false is the default value for bools, but it is the expected result in this case. 
```suggestion
		validatePaste(t, *u)
		if u.Listed != false {
			t.Error("Unlisted paste should have Listed value 'false'.")
		}
```

ejstreet commented on omglol/paste_test.go:
```diff
@@ -0,0 +1,90 @@
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+}
```
Rinse and repeat with `Listed` set to `true`.
```suggestion
        sleep()
        listedTitle := "listedtest"+RunUID
    	listedPaste := NewPaste(listedTitle, "example paste content", true)

	err = c.CreatePaste(testOwnedDomain, *listedPaste)
	if err != nil {
		t.Errorf(err.Error())
	}

	l, err := c.GetPaste(testOwnedDomain, listedTitle)
	if err != nil {
		t.Errorf(err.Error())
	}

	if l != nil {
		t.Log(l.String())
		validatePaste(t, *u)
		if u.Listed != true {
			t.Error("Listed paste should have Listed value 'true'.")
		}
	} else {
		t.Error("GetPaste returned 'nil' when retrieving listed paste.")
	}

	err = c.DeletePaste(testOwnedDomain, title)
	if err != nil {
		t.Errorf(err.Error())
	}
}
```

ejstreet commented on omglol/paste_test.go:
```diff
@@ -0,0 +1,90 @@
+		t.Log(u.String())
+		validatePaste(t, *u)
+	} else {
+		t.Error("GetPaste returned 'nil' when retrieving paste.")
```
```suggestion
		t.Error("GetPaste returned 'nil' when retrieving unlisted paste.")
```

ejstreet commented on omglol/paste_test.go:
```diff
@@ -0,0 +1,90 @@
+		t.Errorf(err.Error())
+	}
+
+	title := "test" + RunUID
```
```suggestion
	unlistedTitle := "unlistedtest" + RunUID
```

ejstreet reviewed (changes requested):
Mark, thank you so much for this! You've done a super job, the code keeps the style of the rest of the repo perfectly!

The only thing that really needs addressing is adding the `listed` attribute, which isn't documented but does appear when querying the current version of the API. It should just be a case of applying suggestions, but I haven't tested them, so not 100% sure they will work 😅

mcornick committed:
Updates to handle Listed attribute and others

Co-authored-by: Elliott Street <ejstreet@users.noreply.github.com>
Signed-off-by: Mark Cornick <mcornick@mcornick.com>

mcornick commented:
OK, I've incorporated your suggested changes, plus a few fixes to make all the tests pass the new code.

ejstreet committed:
Fix ListPaste

I realised from my local testing that the response couldn't use the Paste type, since the response returns an *int rather than a bool.

I also found that using the loop assignment method meant entries were being replicated (reusing the same pointer). Using the NewPaste function in the loop fixes this

ejstreet commented:
Had one fix to add, but I think it's good to go now!
